packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts

```diff
@@ -0,0 +1,302 @@
+    const sources: SourceFile[] = [];
+    const names: string[] = [];
+
+    // Ensure a mapping line array for each line in the generated source.
+    const mappings: SourceMapMappings = new Array(this.lines.length);
+    for (let i = 0; i < mappings.length; i++) {
+      mappings[i] = [];
+    }
+
+    for (const mapping of this.flattenedMappings) {
+      const mappingLine = mappings[mapping.generatedSegment.line];
+      const sourceIndex = findIndexOrAdd(sources, mapping.originalSource);
+      const mappingArray: SourceMapSegment = [
+        mapping.generatedSegment.column,
+        sourceIndex,
+        mapping.originalSegment.line,
+        mapping.originalSegment.column,
+      ];
+      if (mapping.name !== undefined) {
+        const nameIndex = findIndexOrAdd(names, mapping.name);
+        mappingArray.push(nameIndex);
+      }
+      mappingLine.push(mappingArray);
+    }
+
+    const sourcePathDir = dirname(this.sourcePath);
+    const sourceMap: RawSourceMap = {
+      version: 3,
+      file: relative(sourcePathDir, this.sourcePath),
+      sources: sources.map(sf => relative(sourcePathDir, sf.sourcePath)), names,
+      mappings: encode(mappings),
+      sourcesContent: sources.map(sf => sf.contents),
+    };
+    return sourceMap;
+  }
+
+  /**
+   * Flatten the parsed mappings for this source file, so that all the mappings are to pure original
+   * source files with no transitive source maps.
+   */
+  private flattenMappings(): Mapping[] {
+    const mappings = parseMappings(this.rawMap, this.sources);
+    const originalSegments = extractOriginalSegments(mappings);
+    const flattenedMappings: Mapping[] = [];
+    for (let mappingIndex = 0; mappingIndex < mappings.length; mappingIndex++) {
+      const aToBmapping = mappings[mappingIndex];
+      const bSource = aToBmapping.originalSource;
+      if (bSource.flattenedMappings.length === 0) {
+        // The b source file has no mappings of its own (i.e. it a pure original file)
+        // so just use the mapping as-is.
+        flattenedMappings.push(aToBmapping);
+        continue;
+      }
+
+      // The `incomingStart` and `incomingEnd` are the `SegmentMarker`s in `B` that represent the
+      // section of `B` source file that is being mapped to by the current `aToBmapping`.
+      //
+      // For example, consider the mappings from A to B:
+      //
+      // src A   src B     mapping
+      //
+      //   a ----- a       [0, 0]
+      //   b       b
+      //   f -  /- c       [4, 2]
+      //   g  \ /  d
+      //   c -/\   e
+      //   d    \- f       [2, 5]
+      //   e
+      //
+      // For mapping [0,0] the incoming start and end are 0 and 2 (i.e. the range a, b, c)
+      // For mapping [4,2] the incoming start and end are 2 and 5 (i.e. the range c, d, e, f)
+      //
+      const incomingStart = aToBmapping.originalSegment;
+      const incomingEnd = originalSegments[originalSegments.indexOf(incomingStart) + 1];
```

Comment:
nit: since this can be an out of bound read, `incomingEnd` can be `undefined` here (which you have dealt with in all places, so that's all good). AFAIK TypeScript will infer the type of `incomingEnd` to be `SegmentMarker`, so including `undefined` explicitly in the type makes this a bit more safe:

```suggestion
      const incomingEnd: SegmentMarker|undefined = originalSegments[originalSegments.indexOf(incomingStart) + 1];
```

Reply:
I changed the code to

```
      const incomingEnd = incomingEndIndex < originalSegments.length ?
          originalSegments[incomingEndIndex] :
          undefined;
```

to avoid the out-of-bounds lookup - and this also had the benefit of adding the `| undefined` to the type automatically.